### PR TITLE
Rust Backend Rewrite

### PR DESCRIPTION
**Pull request by joeoneil**
Rewrote the backend in Rust to allow for more stability in the API interface and allow expansion of the backend to communicate with games for saving / loading, as well as gatekeeper integration down the line.

### CODE REVIEW

File: onboard/backend/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+            PRODUCTION = prod;
+        }
+    }
+}
+
+/**
+ * A game from the Devcade API
+ */
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+// TODO Update struct to match new schema
```
Comment (MTFT-Games) [inline onboard/backend/src/lib.rs]: Does this not need to be done prior to merge?

**Review by MTFT-Games — APPROVED** [review]
Im tired and dont know rust but if its tested and runs it looks vaguely alright to me conceptually.

**Comment by WillNilges** [comment]
Is there documentation (compilation/installation guide, environment variable sample file, etc)?

**Comment by WillNilges** [comment]
Got it running for a little bit. I had to use the `.env.template` inside the `onboard/` directory and fill it out. It needed some log file paths, resolution, and a directory, presumably to store game data in. Now it's crashing with cryptic messages
```
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  130 (MIT-SHM)
  Minor opcode of failed request:  3 (X_ShmPutImage)
  Value in failed request:  0x340
  Serial number of failed request:  389
  Current serial number in output stream:  390
```

Can't launch any games. Frontend always returns an error, but doesn't crash (weird UI bugs though). Seeing a lot of panics on the backend.

```
thread '<unnamed>' panicked at 'failed to execute child: Os { code: 2, kind: NotFound, message:
"No such file or directory" }', src/api/mod.rs:361:46
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Perhaps there's some busted path. Probably my fault 🙃 

image

File: onboard/backend/src/lib.rs (hunk 1 above)
Comment (joeoneil) [reply onboard/backend/src/lib.rs]: Forgot to remove this. The definition for devcade game currently lives in api/schema.rs, and is updated to the new format.

**Review by MTFT-Games — CHANGES_REQUESTED** [review]
Actually, imma make you update the readme first

File: README.md [hunk 2]
```diff
@@ -7,12 +7,17 @@ Run the `update_onboard.sh` script located in HACKING
 
 ## Building (manual)
 
-To build and run on the Idiot, do the following from `/onboard`:
+To build and run on the Idiot, do the following from `./onboard/frontend`:
```
Comment (MTFT-Games) [inline README.md]: We are moving away from using the 'idiot' terminology. Use DCU (devcade compute unit) instead

File: README.md [hunk 3]
```diff
@@ -7,12 +7,17 @@ Run the `update_onboard.sh` script located in HACKING
...
 ```
 dotnet publish -c Release -r linux-x64 --no-self-contained
 ```
+And the following fron `./onboard/backend`:
+```
+cargo build --release --target x86_64-unknown-linux-gnu
+```
 
-To put it on the Idiot, compress the `publish` folder located at `\Devcade-onboard\onboard\bin\Release\netcoreapp3.1\linux-x64` and `scp` that to the Idiot.
+To put it on the Idiot, compress the `publish` folder located at `./onboard/frontend/bin/Release/netcoreapp3.1/linux-x64` and `scp` that to the Idiot.
```
Comment (MTFT-Games) [inline README.md]: We are moving away from using the 'idiot' terminology. Use DCU (devcade compute unit) instead

File: README.md [hunk 4]
```diff
@@ -7,12 +7,17 @@ Run the `update_onboard.sh` script located in HACKING
...
+You'll also want to `scp` `./onboard/backend/target/release` to the Idiot.
```
Comment (MTFT-Games) [inline README.md]: We are moving away from using the 'idiot' terminology. Use DCU (devcade compute unit) instead

File: onboard/.env.template [hunk 5]
```diff
@@ -1,3 +1,7 @@
+# This file should be copies to .env and filled with the correct values
```
Comment (MTFT-Games) [inline onboard/.env.template]: This should be added to readme .env section

**Review by MTFT-Games — CHANGES_REQUESTED** [review]
specific readme updates

**Pull request closed without merging**